lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+    };
+
+    llvm::SmallMapVector<mlir::Operation *, AccessType, 8> gpuBufferAllocs;
+    llvm::SmallMapVector<unsigned, AccessType, 8> gpuBufferParams;
+    llvm::SmallMapVector<mlir::Operation *, AccessType, 8>
+        gpuGetMemrefGlobalParams;
+    auto &aliases = getAnalysis<mlir::BufferViewFlowAnalysis>();
+
+    auto getMemref = [](mlir::Operation *op)
+        -> llvm::Optional<mlir::SmallVector<mlir::Value, 4>> {
+      if (auto load = mlir::dyn_cast<mlir::memref::LoadOp>(op)) {
+        return {{load.memref()}};
+      } else if (auto store = mlir::dyn_cast<mlir::memref::StoreOp>(op)) {
+        return {{store.memref()}};
+      } else if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        mlir::SmallVector<mlir::Value, 4> ret;
+        for (auto arg : call.operands()) {
+          if (arg.getType().isa<mlir::MemRefType>())
+            ret.emplace_back(arg);
+        }
+        return std::move(ret);
+      } else {
+        op->emitError("Uhhandled mem op in gpu region");
+        return llvm::None;
+      }
+    };
+
+    auto scfDialect = getContext().getOrLoadDialect<mlir::scf::SCFDialect>();
+
+    auto hasMemAccess = [](mlir::Operation *op) -> bool {
+      if (auto memInterface =
+              mlir::dyn_cast<mlir::MemoryEffectOpInterface>(op)) {
+        if (memInterface.hasEffect<mlir::MemoryEffects::Read>() ||
+            memInterface.hasEffect<mlir::MemoryEffects::Write>())
+          return true;
+      }
+      if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        for (auto arg : call.operands()) {
+          if (arg.getType().isa<mlir::MemRefType>())
+            return true;
+        }
+      }
+      return false;
+    };
+
+    // Traverse through all the memory access ops under GPU launch Op
+    // and add device memory allocation appropriately.
+    if (func.walk([&](mlir::Operation *op) {
+              if (!op->getParentOfType<mlir::gpu::LaunchOp>())
+                return mlir::WalkResult::advance();
+
+              if (!hasMemAccess(op))
+                return mlir::WalkResult::advance();
+
+              auto memref = getMemref(op);
+              if (!memref)
+                return mlir::WalkResult::interrupt();
+
+              for (auto mem : *memref) {
+                while (auto parentView =
+                           mem.getDefiningOp<mlir::ViewLikeOpInterface>())
+                  mem = parentView.getViewSource();
+
+                for (auto alias : aliases.resolve(mem)) {
+                  auto op = alias.getDefiningOp();
+                  if (op) {
+                    if (mlir::isa<mlir::memref::GetGlobalOp>(op)) {
+                      gpuGetMemrefGlobalParams.insert({op, {}});
+                      continue;
+                    }
+                    if (op->getDialect() == scfDialect ||
+                        mlir::isa<mlir::ViewLikeOpInterface>(op))
+                      continue;
+                    if (mlir::isa<mlir::memref::AllocOp>(op)) {
```

Comment:
What about handling variants of Allocop? 
[memref.alloc (::mlir::memref::AllocOp)](https://mlir.llvm.org/docs/Dialects/MemRef/#memrefalloc-mlirmemrefallocop)
[memref.alloca (::mlir::memref::AllocaOp)](https://mlir.llvm.org/docs/Dialects/MemRef/#memrefalloca-mlirmemrefallocaop)
[memref.alloca_scope (::mlir::memref::AllocaScopeOp)](https://mlir.llvm.org/docs/Dialects/MemRef/#memrefalloca_scope-mlirmemrefallocascopeop)
[memref.alloca_scope.return (::mlir::memref::AllocaScopeReturnOp)](https://mlir.llvm.org/docs/Dialects/MemRef/#memrefalloca_scopereturn-mlirmemrefallocascopereturnop)

Reply:
these are not handled as of now in this PR, only memref::allocOp is handled

Reply:
Are these cases important to handle?  Or we choose to tell users upfront that we are not handling this?  I would prefer the functionality to cover broad cases if the efforts are reasonable. 

We don't want to find out these corner cases.  We should either implement them or documents them. 

It is fine for this PR not to implementation. But would be good to know the limitation and the plan.

Reply:
added a limitations section in the doc description